File: p35/p35_taguchi.rb
array = []
File.open('./p35_case3.in') { |file| file.each_line { |x| array << x.chomp } }
@yard = array[2..-1].map { |g| g.split('') }

def dfs(x, y)
  @yard[x][y] = '.'
  for dx in -1..1
    for dy in -1..1
      nx = x + dx
      ny = y + dy
      dfs(nx, ny) if 0 <= nx && nx < @yard.length && 0 <= ny && ny < @yard.first.length && @yard[nx][ny] == 'W'
    end
  end
end

@count = 0
for x in 0..(@yard.length - 1) do
  for y in 0..(@yard.first.length - 1) do
    if @yard[x][y] == 'W'
      dfs(x, y)
      @count += 1
    end
  end
end

p @count

File: p35/p35_taguchi.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

var file, _ = os.Open(`./p35_case3.in`)
var sc = bufio.NewScanner(file)

func nextLine() string {
	sc.Scan()
	return sc.Text()
}

var n, _ = strconv.Atoi(nextLine())
var m, _ = strconv.Atoi(nextLine())

func splitYard() [][]string {
	y := make([]string, n)
	a := make([][]string, 0)
	for i := 0; i < n; i++ {
		y = strings.Split(nextLine(), "")
		a = append(a, y)
	}
	return a
}

var yard = splitYard()

func dfs(x int, y int) {
	yard[x][y] = "."
	for dx := -1; dx <= 1; dx++ {
		for dy := -1; dy <= 1; dy++ {
			nx := x + dx
			ny := y + dy
			if 0 <= nx && nx < n && 0 <= ny && ny < m && yard[nx][ny] == "W" {
				dfs(nx, ny)
			}
		}
	}
}

var count = 0

func main() {
	for x := 0; x < n; x++ {
		for y := 0; y < m; y++ {
			if yard[x][y] == "W" {
				dfs(x, y)
				count++
			}
		}
	}
	fmt.Print(count)
}
